stellar/txobserver/main.go

```diff
@@ -48,111 +59,149 @@ LedgerLoop:
+
+	defer func() {
+		_ = o.store.Session.Rollback()
+	}()
+
+	// Get latest list of hashes to observe
+	outgoingBridgeTransactions, err := o.store.GetOutgoingStellarTransactions(context.TODO())
+	if err != nil {
+		return errors.Wrap(err, "error getting outgoing bridge transactions")
+	}
+
+	// If no transactions to observe, skip to next ledger.
+	if len(outgoingBridgeTransactions) == 0 {
+		o.log.WithField("sequence", o.ledgerSequence).Info("No outgoing bridge transactions, skiping to next ledger")
+		return nil
+	}
+
+	// Transform to hash->tx map for faster lookups
+	outgoingBridgeTransactionsHashMap := make(map[string]store.OutgoingStellarTransaction)
+	for _, tx := range outgoingBridgeTransactions {
+		outgoingBridgeTransactionsHashMap[tx.Hash] = tx
+	}
 
-		// If no transactions to observe, skip to next ledger.
-		if len(outgoingBridgeTransactions) == 0 {
-			o.log.WithField("sequence", o.ledgerSequence).Info("No outgoing bridge transactions, skiping to next ledger")
-			o.ledgerSequence++
-			continue
+	// Transform to source->seqnum->tx map for faster lookups
+	outgoingBridgeTransactionsSourceMap := make(map[string]map[int64]store.OutgoingStellarTransaction)
+	for _, tx := range outgoingBridgeTransactions {
+		if outgoingBridgeTransactionsSourceMap[tx.Source] == nil {
+			outgoingBridgeTransactionsSourceMap[tx.Source] = make(map[int64]store.OutgoingStellarTransaction)
 		}
+		outgoingBridgeTransactionsSourceMap[tx.Source][tx.SequenceNumber] = tx
+	}
 
-		// Transform to map for faster lookups
-		outgoingBridgeTransactionsMap := make(map[string]store.OutgoingStellarTransaction)
-		for _, tx := range outgoingBridgeTransactions {
-			outgoingBridgeTransactionsMap[tx.Hash] = tx
+	// Process transactions
+	cursor := ""
+	for {
+		txs, err := o.client.Transactions(horizonclient.TransactionRequest{
+			ForLedger:     uint(o.ledgerSequence),
+			Cursor:        cursor,
+			Limit:         200,
+			IncludeFailed: true,
+		})
+		if err != nil {
+			return errors.Wrap(err, "error getting transactions")
 		}
 
-		// Process transactions
-		cursor := ""
-		for {
-			txs, err := o.client.Transactions(horizonclient.TransactionRequest{
-				ForLedger:     uint(o.ledgerSequence),
-				Cursor:        cursor,
-				Limit:         200,
-				IncludeFailed: true,
-			})
-			if err != nil {
-				o.log.WithFields(slog.F{
-					"error":  err,
-					"ledger": o.ledgerSequence,
-					"cursor": cursor,
-				}).Error("Error getting transactions")
-				time.Sleep(time.Second)
-				continue LedgerLoop
-			}
+		if len(txs.Embedded.Records) == 0 {
+			break
+		}
 
-			if len(txs.Embedded.Records) == 0 {
-				break
+		for _, tx := range txs.Embedded.Records {
+			otx, hashExists := outgoingBridgeTransactionsHashMap[tx.Hash]
+			if hashExists {
+				if tx.Successful {
+					otx.State = store.SuccessState
+					// TODO [important] update IncomingEthereumTransaction state
+					// as withdrawn so it's not possible to cancel it.
+				} else {
+					otx.State = store.FailedState
```

Comment:
Do we really need a failed state or would the expired state be sufficient? I'm presuming that a failed tx nevertheless consumes its sequence number, and so failed is a special case of expired.

Reply:
> so failed is a special case of expired

Exactly. I leaning towards leaving the `FailedState` in case in the future we want to show some status update to the client. `FailedState` means there was some kind of issue with their account config (like missing seqnum) instead of just passage of time / invalidation by other tx.